### PR TITLE
Sin, cos, sincos tan, and more rem_pio2

### PR DESCRIPTION
**Pull request by pkofod**
This PR builds on #22603 . I have included the changes from that PR as a first commit here until it gets merged.

This PR adds a `rem_pio2` method for `Float32` as well, and gets rid of openlibm calls for `sin`, `cos` and `sincos` for `Float32` and `Float64`.

Performance seems to regress a bit, and I'm not entirely sure why, but still investigating (https://github.com/JuliaCI/BaseBenchmarks.jl/pull/81 is related).

### CODE REVIEW

**Comment by pkofod** [comment]
I'm seeing some performance improvements here, but I'm not going to conclude anything until BaseBenchmarks is able to run the perf tests.

File: base/special/trig.jl [hunk 1]
```diff
@@ -20,62 +20,182 @@ end
...
+        return co, -si
+    elseif n == 2
+        return -si, -co
+    else
+        return -co, si
+    end
+end
+
+# There's no need to write specialized kernels, as inlining takes care of remo-
+# ing superfluous calculations.
```
Comment (Sacha0) [inline base/special/trig.jl]: "remo- ing"? :)

File: base/special/trig.jl [hunk 2]
```diff
@@ -20,62 +20,310 @@ end
...
+    #		       = 1 - 2*(tan(z) - (tan(z)^2)/(1+tan(z)))
+
+    # do the same "if 0 thing" as for ylo of sin cos
+
+    # The "low word" of the double-double approximation to π/4
+    pio4lo = 3.06161699786838301793e-17
+    yhi = y.hi
+    ylo = y.lo
+
+	if abs(yhi) >= 0.6744
```
Comment (tkelman) [inline base/special/trig.jl]: spaces, not tabs

**Comment by pkofod** [comment]
Is it possible for someone to run the `sin` and `cos` benchmark groups on nanosoldier? I didn't include `tan` when I made that PR, but I'll include them so it can be run before this is merged at some point. (maybe you @simonbyrne  ? :))

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL,  vs = ":master")`

File: base/math.jl [hunk 3]
```diff
@@ -436,27 +436,18 @@ julia> log1p(0)
...
-@inline function sincos(x)
-    res = Base.FastMath.sincos_fast(x)
-    if (isnan(res[1]) | isnan(res[2])) & !isnan(x)
-        throw(DomainError(x, "NaN result for non-NaN input."))
-    end
-    return res
-end
+@inline sin(x::Real) = sin(float(x))
+@inline cos(x::Real) = cos(float(x))
+@inline sincos(x::Real) = cos(float(x))
```
Comment (giordano) [inline base/math.jl]: What?

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22824#issuecomment-318483787):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ef4b369d80331ebf1339681fec73995b3fa304b0/5290a67_vs_85a2555/report.md)
cc @ararslan

**Comment by ararslan** [comment]
Nanosoldier has been having issues lately due to a bug in JLD (or maybe HDF5).

File: base/math.jl (hunk 3 above)
Comment (pkofod) [inline base/math.jl]: Yeah that last line is missing a sin on the rhs :)

File: base/special/trig.jl [hunk 4]
```diff
@@ -20,65 +20,351 @@ end
...
+        return Py
+    end
+
+    return -1.0/Py
+end
+
 # fallback methods
 sin_kernel(x::Real) = sin(x)
 cos_kernel(x::Real) = cos(x)
+sincos_kernel(x::Real) = cos(x)
```
Comment (giordano) [inline base/special/trig.jl]: I guess this is also wrong:
```julia
julia> sincos(1.0)
(0.8414709848078965, 0.5403023058681398)

julia> sincos(0.1)
0.9950041652780258
```

File: base/special/trig.jl [hunk 5]
```diff
@@ -26,7 +26,7 @@ end
 ### sin methods
 @inline function sin{T<:AbstractFloat}(x::T)
     if abs(x) < T(pi)/4 #|x| ~<= pi/4, no need for reduction
-        if x < sqrt(eps(T))
+        if ax < sqrt(eps(T))
```
Comment (KristofferC) [inline base/special/trig.jl]: Where is `ax` defined? Maybe I'm missing something obvious.

Comment (pkofod) [inline base/special/trig.jl]: It isn't. You aren't. :)

File: base/math.jl [hunk 6]
```diff
@@ -826,6 +817,7 @@ julia> rem2pi(7pi/4, RoundDown)
 5.497787143782138
 ```
 """
+
```
Comment (tkelman) [inline base/math.jl]: preferable not to have any empty lines between a docstring and the signature it's documenting

**Comment by ararslan** [comment]
The Nanosoldier is all better now. @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22824#issuecomment-322276164) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/56c80f621008d5520a52810be434abfc906f15bb/89f3e17_vs_a62160e/report.md). cc @ararslan

**Comment by pkofod** [comment]
Cleaned up the commits, and I'll try to make this a finished PR. `tan` has a bug that I will sort out tomorrow, and `rem_pio2` for `Float32` needs to be tested.

Besides testing some values in the branches, is there anything I should do here? Could choose some random values as well.

**Comment by pkofod** [comment]
I'm going to add tests for the `rem_pio2(x::Float32)` method, but besides that, I would love some comments should you have any. cc @simonbyrne @musm

**Comment by simonbyrne** [comment]
Looks pretty good, once we have the `tan` benchmarks in place we can merge.

**Comment by ararslan** [comment]
Fresh off the Nanosoldier: @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22824#issuecomment-332101675) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/76d671cabc43664040d0c1e1094d7c0f2acbec0a/2b968df_vs_da43160/report.md). cc @ararslan

**Comment by fredrikekre** [comment]
:white_check_mark:

**Comment by ararslan** [comment]
Oh man, those scalar benchmarks. Excellent work, @pkofod!

**Comment by simonbyrne** [comment]
That looks great! The only relevant ones with reduced performance seem to be:

["scalar","cos",("no reduction","zero","Float32")] | 1.29 (15%) ❌ | 1.00 (1%)
["scalar","cos",("no reduction","zero","Float64")] | 1.29 (15%) ❌ | 1.00 (1%)


Probably not that big of a deal compared with the better performance of the rest. Does openlibm have an early-out check for zero?

**Comment by pkofod** [comment]
> Probably not that big of a deal compared with the better performance of the rest. Does openlibm have an early-out check for zero?

Compare: https://github.com/JuliaLang/julia/pull/22824/files#diff-e0ba4eb753f5036415cadf8a84ad0a5cR31

to openlibm:
https://github.com/JuliaLang/openlibm/blob/ddf649beabe0afe92e1b31739b2fa520cd7e239f/src/s_sin.c#L66

**Comment by jebej** [comment]
> Probably not that big of a deal compared with the better performance of the rest. Does openlibm have an early-out check for zero?
> Compare: https://github.com/JuliaLang/julia/pull/22824/files#diff-e0ba4eb753f5036415cadf8a84ad0a5cR31
> 
> to openlibm:
> https://github.com/JuliaLang/openlibm/blob/ddf649beabe0afe92e1b31739b2fa520cd7e239f/src/s_sin.c#L66

Is the `sqrt(eps(T))` computed at compile time?

**Comment by pkofod** [comment]
> Is the sqrt(eps(T)) computed at compile time?

Yes
```julia
/                   |  x86_64-pc-linux-gnu

julia> function f(x::T)
       sqrt(eps(T))
       end
ERROR: UndefVarError: T not defined

julia> function f(x::T) where T
           sqrt(eps(T))
       end
f (generic function with 1 method)

julia> f(0.1)
1.4901161193847656e-8

julia> @code_warntype f(0.1)
Variables:
  #self#::#f
  x::Float64

Body:
  begin 
      return (Base.Math.sqrt_llvm)(2.220446049250313e-16)::Float64
  end::Float64

julia> @code_native f(0.1)
	.text
Filename: REPL[2]
	pushq	%rbp
	movq	%rsp, %rbp
	movabsq	$140587044811840, %rax  # imm = 0x7FDCF8DDB440
Source line: 2
	movsd	(%rax), %xmm0           # xmm0 = mem[0],zero
	popq	%rbp
	retq
	nopw	%cs:(%rax,%rax)

julia> @code_llvm f(0.1)

define double @julia_f_60749(double) #0 !dbg !5 {
top:
  ret double 0x3E50000000000000
}
```

**Comment by pkofod** [comment]
Oh! I read your message incorrectly @simonbyrne , I thought we were in `sin`! But we're in `cos`. Then yes, it does seem like I missed a return `1.0` if we have small values of `x`!
https://github.com/JuliaLang/julia/pull/22824/files#diff-e0ba4eb753f5036415cadf8a84ad0a5cR97
vs 
https://github.com/JuliaLang/openlibm/blob/ddf649beabe0afe92e1b31739b2fa520cd7e239f/src/s_cos.c#L66

**Comment by pkofod** [comment]
Let's try nanosoldier again!

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks("scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22824#issuecomment-333750332) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/07d59215ca5ca69b0b2ac5061afe4d53f73c1964/fee69fc_vs_6159bcb/report.md). cc @ararslan

**Comment by KristofferC** [comment]
Very nice benchmarks.

File: base/special/trig.jl [hunk 7]
```diff
@@ -22,65 +22,328 @@ end
...
+@inline function sin_kernel(y::DoubleFloat64)
+    y² = y.hi*y.hi
+    y⁴ = y²*y²
+    r  = @horner(y², DS2, DS3, DS4) + y²*y⁴*@horner(y², DS5, DS6)
+    y³ = y²*y.hi
+    y.hi-((y²*(0.5*y.lo-y³*r)-y.lo)-y³*DS1)
+end
+@inline function sin_kernel(y::Float64)
+    y² =  y*y
+    y⁴ =  y²*y²
```
Comment (KristofferC) [inline base/special/trig.jl]: It is a little bit confusing to me why y^2 *y^2 is called y.

Comment (pkofod) [reply base/special/trig.jl]: there's  a small 4 super script there... maybe I should ditch the unicode?

Comment (KristofferC) [reply base/special/trig.jl]: Oh, it was completely missing on my phone.

Comment (pkofod) [reply base/special/trig.jl]: I will say, that even on my computer, it's very small here on github :) It's just easier to follow the code than names like `x` `y` `z` `w` for all kinds of powers of `y` imo, but it's of course a problem if it's hard to read other places than my own editor :)

Comment (KristofferC) [reply base/special/trig.jl]: I'm fine with keeping it.

**Comment by giordano** [comment]
`sincos` is type unstable:
```julia
julia> @inferred sincos(1.0)
ERROR: return type Tuple{Float64,Float64} does not match inferred return type Union{Float64, Tuple{Float64,Float64}}
Stacktrace:
 [1] error(::String) at ./error.jl:33
```

File: base/special/trig.jl [hunk 8]
```diff
@@ -22,65 +22,328 @@ end
...
+Simultaneously compute the sine and cosine of `x`, where the `x` is in radians.
+"""
+@inline function sincos(x::T) where T<:Union{Float32, Float64}
+    if abs(x) < T(pi)/4
+        if x == zero(T)
+            return x, one(T)
+        end
+        return sincos_kernel(x)
+    elseif isnan(x)
+        return T(NaN)
```
Comment (giordano) [inline base/special/trig.jl]: This is the issue, should return `T(NaN), T(NaN)`.

Comment (pkofod) [reply base/special/trig.jl]: Thanks for catching this!

**Comment by KristofferC** [comment]
Are there no benchmarks for `sincos` then?

File: base/special/trig.jl (hunk 8 above)
Comment (giordano) [reply base/special/trig.jl]: You're welcome, you made an impressive work!  Is this a faithful translation from openlibm or are there other improvements?  The speedup is awesome.

**Comment by pkofod** [comment]
> Are there no benchmarks for sincos then?

It appears that I didn't add any, no https://github.com/JuliaCI/BaseBenchmarks.jl/blob/5b733a2801e9bdb69c32913af802afefa3d78c31/src/scalar/ScalarBenchmarks.jl

**Comment by KristofferC** [comment]
Might be good to benchmark to explicitly test the

```
# There's no need to write specialized kernels, as inlining takes care of remo-
 # ving superfluous calculations.
```

comment.

**Comment by pkofod** [comment]
> Might be good to benchmark to explicitly test the
> 
> # There's no need to write specialized kernels, as inlining takes care of remo-
> # ving superfluous calculations.
> 
> comment.

I will. It's also important to keep track of this not regressing in the future.

**Comment by pkofod** [comment]
Once https://github.com/JuliaCI/BaseBenchmarks.jl/pull/126 is merged and benchmarked I believe this should be gtg.

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks("scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22824#issuecomment-334658542) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/4edbcf239b928f56496a6b93babcddbd06bedcd9/42136fc_vs_7504a38/report.md). cc @ararslan

**Comment by ararslan** [comment]
That's a whole lot of ✅.

**Comment by pkofod** [comment]
Yup, the `sincos` inlining seems to work just fine.

Though, it does make me think that the `atan2` "regressions" in the `atan` PR are spurious, since this PR should do nothing to atan2 when `atan` isn't even merged yet, and therefore can't depend on `sin`, `cos`, or `tan`.

**Review by simonbyrne — APPROVED** [review]

**Pull request merged**